// @fivethree/billy-plugin-git/testi/immoapp/src/app/app-routing.module.ts
import { AuthGuard } from './guards/auth.guard';
import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

const routes: Routes = [
  {
    path: '',
    loadChildren: './tabs/tabs.module#TabsPageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'login',
    loadChildren: './pages/login/login.module#LoginPageModule'
  },
  {
    path: 'properties/:id',
    loadChildren: './pages/property/property.module#PropertyPageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'check',
    loadChildren: './pages/check/check.module#CheckPageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'inspect/:id',
    loadChildren: './pages/inspect/inspect.module#InspectPageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'inspection/details/:id',
    loadChildren: './pages/inspection-detail/inspection-detail.module#InspectionDetailPageModule',
    canActivate: [AuthGuard]
  },

  {
    path: 'messages/:id',
    loadChildren: './pages/message/message.module#MessagePageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'password/reset',
    loadChildren: './pages/password-reset/password-reset.module#PasswordResetPageModule',
  },
  {
    path: 'password/change',
    loadChildren: './pages/password-change/password-change.module#PasswordChangePageModule',
    canActivate: [AuthGuard]
  },
  {
    path: 'imprint',
    loadChildren: './pages/imprint/imprint.module#ImprintPageModule'
  },
  {
    path: 'privacy',
    loadChildren: './pages/privacy/privacy.module#PrivacyPageModule'
  },
  {
    path: 'user/detail',
    loadChildren: './pages/user-detail/user-detail.module#UserDetailPageModule',
    canActivate: [AuthGuard]
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
})
export class AppRoutingModule { }
