// @fivethree/billy-plugin-git/testi/management-app/src/app/pages/home/home.page.ts
import { DataService } from '@services/data.service';
import { Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';
import { Checklist, Property, User, Inspection } from 'shared';
import { Router } from '@angular/router';
import { map, filter, flatMap } from 'rxjs/operators';
import { Store } from '@services/store/store.service';

@Component({
  selector: 'app-home',
  templateUrl: 'home.page.html',
  styleUrls: ['home.page.scss'],
})
export class HomePage {



  adminchecklists: Observable<Checklist[]>;
  inspections: Observable<Inspection[]>;
  checklists: Observable<Checklist[]>;
  properties: Observable<Property[]>;
  users: Observable<User[]>;
  verwalter: Observable<User[]>;
  user: Observable<User>;


  constructor(public data: DataService, public router: Router, private store: Store) {
  }

  ionViewWillEnter(): void {
    this.user = this.store.select('user');

    this.properties = this.user
      .pipe(
        filter((user: User) => user.role !== 'ROLE_ADMIN'),
        flatMap(() => this.data.loadProperties())
      );

    this.checklists = this.user
      .pipe(
        filter((user: User) => user.role !== 'ROLE_ADMIN'),
        flatMap(() => this.data.loadChecklists())
      );

    this.adminchecklists = this.user
      .pipe(
        filter((user: User) => user.role === 'ROLE_ADMIN'),
        flatMap(() => this.data.loadAdminChecklists())
      );


    this.users = this.user
      .pipe(
        filter((user: User) => user.role !== 'ROLE_ADMIN'),
        flatMap(() => this.data.loadUsers())
      );

    this.inspections = this.user
      .pipe(
        filter((user: User) => user.role !== 'ROLE_ADMIN'),
        flatMap(() => this.data.loadInspections())
      );

    this.verwalter = this.user
      .pipe(
        filter((user: User) => user.role === 'ROLE_ADMIN'),
        flatMap(() => this.data.loadHausverwalter())
      );


  }

  addUser() {
    this.router.navigate(['users/edit']);
  }

  addHausverwalter() {
    this.router.navigate(['users/edit']);
  }

  addProperty() {
    this.router.navigate(['properties/edit']);
  }

  addChecklist() {
    this.router.navigate(['checklists/create']);
  }

  addChecklistTemplate() {
    this.router.navigate(['checklists/create']);
  }

  goToSettings() {
    this.router.navigate(['settings']);
  }

  goToInspections() {
    this.router.navigate(['inspections']);
  }
  goToInspection(id) {
    this.router.navigate(['inspections', id]);
  }

  navigateToUsers(user: User) {
    if (user.role === 'ROLE_ADMIN') {
      this.router.navigate(['hausverwalter']);
    } else {
      this.router.navigate(['users']);
    }
  }
  navigateToUser(id, user) {
    if (user.role === 'ROLE_ADMIN') {
      this.router.navigate(['hausverwalter', id]);
    } else {
      this.router.navigate(['users', id]);
    }
  }
  navigateToChecklists(user) {
    if (user.role === 'ROLE_ADMIN') {
      this.router.navigate(['admin-checklists']);
    } else {
      this.router.navigate(['checklists']);
    }
  }
  navigateToChecklist(id, user) {
    this.router.navigate(['checklists', id]);
  }

}
