// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/components/kriterium/kriterium.component.ts
import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { Kriterien } from 'shared';
import { map, tap } from 'rxjs/operators';
import { DomSanitizer, SafeResourceUrl } from '@angular/platform-browser';
import { ImageService } from '@services/image.service';

@Component({
  selector: 'app-kriterium',
  templateUrl: './kriterium.component.html',
  styleUrls: ['./kriterium.component.scss']
})
export class KriteriumComponent implements OnInit {

  @Input() kriterium: Kriterien;
  @Input() checked = false;
  imageData: SafeResourceUrl;

  @Output() change = new EventEmitter<any>();

  constructor(public imageService: ImageService, 
    private dom: DomSanitizer) { }

  ngOnInit() {
    this.loadImage();
  }

  loadImage() {
    if (this.kriterium.picto) {
      this.imageService.getPictoBase64(this.kriterium.picto.id)
        .pipe(
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
          tap(image => console.log('image', image)),
          tap(image => this.imageData = image),
        ).subscribe(() => {
          console.log('loaded image');
        }, err => console.log('couldnt load image', err));
    }
  }

  onCheckedChange(event) {
    this.checked = event.detail.checked;
  }

}
