// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/inspection-kriterium/inspection-kriterium.component.ts
import { ApiService } from './../../services/api.service';
import { Component, OnInit, Input } from '@angular/core';
import { Inspection, Result, Criteria } from 'shared';
import { SafeResourceUrl, DomSanitizer } from '@angular/platform-browser';
import { map, tap } from 'rxjs/operators';

@Component({
  selector: 'app-inspection-kriterium',
  templateUrl: './inspection-kriterium.component.html',
  styleUrls: ['./inspection-kriterium.component.scss']
})
export class InspectionKriteriumComponent implements OnInit {

  @Input() inspection: Inspection;
  @Input() kriterium: Criteria;
  results: Result[] = [];
  imageData: SafeResourceUrl;

  constructor(private api: ApiService, private dom: DomSanitizer) { }

  ngOnInit() {
    console.log(this.inspection, this.kriterium);
    this.results = this.inspection.results
      .filter(res => res.pruefkriterium.id === this.kriterium.pruefkriterium.id);
    this.loadImage();
  }

  loadImage() {
    if (this.kriterium.pruefkriterium.picto) {
      this.api.getPictoBase64(this.kriterium.pruefkriterium.picto.id, false)
        .pipe(
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
          tap(image => console.log('image', image)),
          tap(image => this.imageData = image),
        ).subscribe(() => {
          console.log('loaded image');
        }, err => console.log('couldnt load image', err));
    }
  }

}
