// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/kriterium-select/kriterium-select.component.ts
import { Component, OnInit, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { Kriterien } from 'shared';
import { ApiService } from '../../services/api.service';
import { map, tap } from 'rxjs/operators';
import { DomSanitizer, SafeResourceUrl } from '@angular/platform-browser';
import { IonCheckbox } from '@ionic/angular';

@Component({
  selector: 'app-kriterium-select',
  templateUrl: './kriterium-select.component.html',
  styleUrls: ['./kriterium-select.component.scss']
})
export class KriteriumSelectComponent implements OnInit {

  @Input() kriterium: Kriterien;
  @Input() checked = false;
  imageData: SafeResourceUrl;

  @Output() change = new EventEmitter<boolean>();

  @ViewChild(IonCheckbox) checkbox: IonCheckbox;

  constructor(public api: ApiService, private dom: DomSanitizer) { }

  ngOnInit() {
    this.loadImage();
  }

  loadImage() {
    if (this.kriterium.picto) {
      this.api.getPictoBase64(this.kriterium.picto.id)
        .pipe(
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
          tap(image => console.log('image', image)),
          tap(image => this.imageData = image),
        ).subscribe(() => {
          console.log('loaded image');
        }, err => console.log('couldnt load image', err));
    }
  }

  onCheckedChange(event) {
    this.checked = event.detail.checked;
    this.change.emit(this.checked);
  }

}
