// @fivethree/billy-plugin-git/testi/immoapp/src/app/pages/password-reset/password-reset.page.ts
import { FivLoadingProgressBar } from '@fivethree/core';
import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, FormControl, Validators, AbstractControl } from '@angular/forms';
import { NavController } from '@ionic/angular';
import { ApiService } from '@services/api.service';
import { NotifyService } from '@services/notify.service';

@Component({
  selector: 'app-password-reset',
  templateUrl: './password-reset.page.html',
  styleUrls: ['./password-reset.page.scss'],
})
export class PasswordResetPage implements OnInit {

  passwordReset: FormGroup;

  constructor(public formbuilder: FormBuilder,
    private api: ApiService,
    private navCtrl: NavController,
    private toast: NotifyService) {
    this.passwordReset = formbuilder.group({
      email: new FormControl('', [Validators.email, Validators.required]),
      email2: new FormControl('', [Validators.email, Validators.required]),
    });
  }

  ngOnInit() {
  }

  get email(): AbstractControl {
    return this.passwordReset.get('email');
  }
  get email2(): AbstractControl {
    return this.passwordReset.get('email2');
  }

  equal() {
    return this.email.value === this.email2.value;
  }

  reset(loading: FivLoadingProgressBar) {
    loading.load();
    this.api.resetOwnPassword(this.email.value)
      .subscribe(() => {
        loading.complete({});
        this.toast.success('Ihr Passwort wurde zurückgesetzt. Bite überprüfen Sie Ihre E-Mails um die Zurücksetzung abzuschließen.');
      }, err => {
        this.toast.error('Ein Fehler ist aufgetreten. Bitte versuchen Sie es später erneut');
        loading.unload();
      });
  }

  resetComplete() {
    this.navCtrl.back();
  }

}
