// @fivethree/billy-plugin-git/testi/immoapp/src/app/pages/login/login.page.ts
import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormControl, Validators, FormGroup, AbstractControl } from '@angular/forms';
import { Store } from '@ngxs/store';
import { Login, LoginSuccess } from '@store/auth/auth.actions';
import { FivLoadingButton } from '@fivethree/core';
import { NotifyError } from '@store/notify/notify.actions';
import { NavController } from '@ionic/angular';

@Component({
  selector: 'app-login',
  templateUrl: './login.page.html',
  styleUrls: ['./login.page.scss'],
})
export class LoginPage implements OnInit {

  loginForm: FormGroup;

  constructor(public formbuilder: FormBuilder,
    private navCtrl: NavController,
    private store: Store) {

    this.loginForm = formbuilder.group({
      email: new FormControl('', [Validators.email, Validators.required]),
      password: new FormControl('', [Validators.minLength(6), Validators.required]),
    });

  }

  ngOnInit() {
  }

  login(button: FivLoadingButton) {
    button.load();
    this.store.dispatch(new Login(this.loginForm.value))
      .subscribe(() => {
        button.complete();
      },
        (error) => {
          button.unload();
          this.store.dispatch(new NotifyError('Entweder ist die E-Mail-Adresse oder das Passwort nicht korrekt.'));
        });
  }

  loginComplete() {
    this.store.dispatch(new LoginSuccess());
  }

  get email(): AbstractControl {
    return this.loginForm.get('email');
  }

  get password(): AbstractControl {
    return this.loginForm.get('password');
  }

  goToPasswordReset() {
    this.navCtrl.navigateForward('/password/reset', { animated: true });
  }
}
