// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/pages/inspection-detail/inspection-detail.page.ts
import { PropertiesState } from '@store/properties/properties.state';
import { ActivatedRoute } from '@angular/router';
import { NavigateBackward } from '@fivethree/ngxs-ionic-router-plugin';
import { Component, OnInit } from '@angular/core';
import { Store } from '@ngxs/store';
import { Observable } from 'rxjs';
import { Inspection, Property } from 'shared';
import { InspectionsState } from '@store/inspections/inspections.state';
import { flatMap } from 'rxjs/operators';

@Component({
  selector: 'app-inspection-detail',
  templateUrl: './inspection-detail.page.html',
  styleUrls: ['./inspection-detail.page.scss'],
})
export class InspectionDetailPage implements OnInit {

  inspection: Observable<Inspection>;
  inspectionId: number;
  property: Observable<Property>;


  constructor(private store: Store, private route: ActivatedRoute) { }

  ngOnInit() {
  }

  ionViewWillEnter() {
    this.inspectionId = +this.route.snapshot.paramMap.get('id');
    this.inspection = this.store.select(InspectionsState.getClosedInspection(this.inspectionId));
    this.property = this.inspection
      .pipe(
        flatMap(inspection => this.store.select(PropertiesState.getProperty(inspection.propertyId)))
      );

  }

  goBack() {
    this.store.dispatch(new NavigateBackward('/tabs/tab2'));
  }

}
