// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/app.module.ts
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { RouteReuseStrategy } from '@angular/router';

import { IonicModule, IonicRouteStrategy } from '@ionic/angular';
import { SplashScreen } from '@ionic-native/splash-screen/ngx';
import { StatusBar } from '@ionic-native/status-bar/ngx';

import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { IonicStorageModule } from '@ionic/storage';
import { IonicConfig } from '@ionic/core';
import { LottieAnimationViewModule } from 'ng-lottie';
import { InAppBrowser } from '@ionic-native/in-app-browser/ngx';
import { NgxsStoreModule } from '@store/store.module';
import { AuthInterceptor } from '@interceptors/auth.interceptor';
import { FCM } from '@ionic-native/fcm/ngx';
import { MarkdownModule } from 'ngx-markdown';
import { Camera } from '@ionic-native/camera/ngx';


const ionicConfig: IonicConfig = { mode: 'md' };

@NgModule({
  declarations: [AppComponent],
  entryComponents: [],
  imports: [BrowserModule,
    HttpClientModule,
    IonicModule.forRoot(ionicConfig),
    IonicStorageModule.forRoot(),
    AppRoutingModule,
    BrowserAnimationsModule,
    LottieAnimationViewModule.forRoot(),
    NgxsStoreModule,
    MarkdownModule.forRoot()
  ],
  providers: [
    StatusBar,
    SplashScreen,
    { provide: RouteReuseStrategy, useClass: IonicRouteStrategy },
    { provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true },
    InAppBrowser,
    FCM,
    Camera
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
