// @fivethree/billy-plugin-git/testi/management-app/src/app/pages/admin-checklists/admin-checklists.page.ts
import { Component, OnInit, ViewChild, OnDestroy, AfterViewInit } from '@angular/core';
import { Observable } from 'rxjs';
import { Checklist } from 'shared';
import { DataService } from '@services/data.service';
import { Router } from '@angular/router';
import { MatTableDataSource } from '@angular/material/table';
import { MatPaginator } from '@angular/material/paginator';
import { MatSort } from '@angular/material/sort';

@Component({
  selector: 'app-admin-checklists',
  templateUrl: './admin-checklists.page.html',
  styleUrls: ['./admin-checklists.page.scss'],
})
export class AdminChecklistsPage implements OnInit, OnDestroy, AfterViewInit {

  public checklists: Observable<Checklist[]>;
  searchInput: String = '';

  dataSource = new MatTableDataSource<Checklist>();
  displayedColumns: string[] = ['name', 'levels'];
  @ViewChild(MatPaginator) paginator: MatPaginator;
  @ViewChild(MatSort) sort: MatSort;

  onRowClicked(event) {
    console.log('row clicked', event);
    this.navigateToChecklist(event.id);
  }

  applySearchFilter(event) {
    this.dataSource.filter = event.detail.value.trim().toLowerCase();
  }

  onSearchbarClose() {
    this.dataSource.filter = '';
  }


  constructor(public data: DataService,
    public router: Router) {

  }

  ngAfterViewInit(): void {
    this.paginator._intl.itemsPerPageLabel = 'Checklisten pro Seite';
    this.paginator._intl.firstPageLabel = 'Erste Seite';
    this.paginator._intl.nextPageLabel = 'Nächste Seite';
    this.paginator._intl.previousPageLabel = 'Vorherige Seite';
    this.paginator._intl.lastPageLabel = 'Letzte Seite';
    this.paginator._intl.getRangeLabel = (page: number, pageSize: number, length: number) => {
      const startIndex = page * pageSize;
      const endIndex = startIndex < length ? Math.min(startIndex + pageSize, length) : startIndex + pageSize;
      return `${startIndex + 1} - ${endIndex} von ${length}`;
    };
    this.dataSource.paginator = this.paginator;
    this.dataSource.sort = this.sort;
  }


  ngOnInit() {
  }
  ionViewWillEnter() {
    this.checklists = this.data.loadAdminChecklists();
    this.checklists.subscribe((checklists => this.dataSource.data = checklists));
  }

  rowSelected(event) {
    if (event.type === 'click') {
      this.navigateToChecklist(event.row.id);
    }
  }

  navigateToChecklist(id) {
    this.router.navigate(['checklists', id]);
  }

  search(event) {
    this.searchInput = event.detail.value;
  }

  addChecklist() {

    this.router.navigate(['checklists/create']);
  }

  ngOnDestroy(): void {
  }



}
