// @fivethree/billy-plugin-git/testi/immoapp/src/app/components/inspection/inspection.component.ts
import { Component, OnInit, Input } from '@angular/core';
import { Inspection, Property } from 'shared';
import { Store } from '@ngxs/store';
import { Observable } from 'rxjs';
import { NavigateForward } from '@fivethree/ngxs-ionic-router-plugin';
import { PropertiesState } from '@store/properties/properties.state';

@Component({
  selector: 'app-inspection',
  templateUrl: './inspection.component.html',
  styleUrls: ['./inspection.component.scss']
})
export class InspectionComponent implements OnInit {

  @Input() inspection: Inspection;
  openCriterias = 0;
  property$: Observable<Property>;

  constructor(private store: Store) { }

  ngOnInit() {
    this.openCriterias = this.inspection.criterias.length - this.inspection.criterias
      .filter(crit => this.inspection.results.some(res => res.pruefkriterium.id === crit.pruefkriterium.id)).length;

    this.property$ = this.store.select(PropertiesState.getProperty(this.inspection.propertyId));
  }

  inspect() {
    this.store.dispatch(new NavigateForward(['/inspect', this.inspection.id]));
  }

  showClosedInspection() {
    this.store.dispatch(new NavigateForward(['/inspection/details', this.inspection.id]));
  }

}
