// @fivethree/billy-plugin-git/immoapps/management-app/src/app/app.component.ts
import { Router } from '@angular/router';
import { User } from 'shared';
import { Store } from './services/store/store.service';
import { TokenService } from './services/token/token.service';
import { Component } from '@angular/core';

import { Platform, NavController } from '@ionic/angular';
import { SplashScreen } from '@ionic-native/splash-screen/ngx';
import { StatusBar } from '@ionic-native/status-bar/ngx';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-root',
  templateUrl: 'app.component.html'
})
export class AppComponent {

  public loggedIn = false;
  public user: Observable<User>;

  constructor(
    private platform: Platform,
    private splashScreen: SplashScreen,
    private statusBar: StatusBar,
    public token: TokenService,
    public navCtrl: NavController,
    public store: Store,
    public router: Router
  ) {
    this.initializeApp();

  }

  initializeApp() {
    this.platform.ready().then(() => {
      this.statusBar.styleDefault();
      this.splashScreen.hide();
      this.setup();
    });
  }
  setup() {
    this.user = this.store.select('user');
    this.token.loggedIn.subscribe(loggedIn => {
      console.log('on login state change', loggedIn);
      this.loggedIn = loggedIn;
    });
  }

  goToUserSettings() {
    this.navCtrl.navigateForward('/settings');
  }

  navigateHome(){
    this.navCtrl.navigateRoot('/');
  }
}
