// @fivethree/billy-plugin-git/immoapps/management-app/src/app/app.module.ts
import { ToastService } from './services/toast.service';
import { FivethreeCoreModule } from '@fivethree/core';
import { ApiService } from './services/api.service';
import { DataService } from './services/data.service';
import { TokenService } from './services/token/token.service';
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { RouteReuseStrategy } from '@angular/router';

import { IonicModule, IonicRouteStrategy } from '@ionic/angular';
import { SplashScreen } from '@ionic-native/splash-screen/ngx';
import { StatusBar } from '@ionic-native/status-bar/ngx';

import { AppComponent } from './app.component';
import { IonicStorageModule } from '@ionic/storage';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { AuthInterceptor } from './services/auth.interceptor';
import { UserService } from './services/user.service';
import { AuthService } from './services/auth.service';
import { Store } from './services/store/store.service';
import { IonicConfig } from '@ionic/core';
import { AppRoutingModule } from './app-routing.module';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { MatIconModule } from '@angular/material/icon';

const ionicConfig: IonicConfig = { mode: 'md' };

@NgModule({
  declarations: [AppComponent],
  entryComponents: [],
  imports: [BrowserModule,
    BrowserAnimationsModule,
    HttpClientModule,
    IonicModule.forRoot(ionicConfig),
    IonicStorageModule.forRoot(),
    AppRoutingModule,
    MatIconModule,
    FivethreeCoreModule,
  ],
  providers: [
    StatusBar,
    SplashScreen,
    { provide: RouteReuseStrategy, useClass: IonicRouteStrategy },
    { provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true },
    TokenService,
    UserService,
    AuthService,
    Store,
    DataService,
    ApiService,
    ToastService

  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
