// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/level-inspector-kriterium/level-inspector-kriterium.component.ts
import { ApiService } from './../../services/api.service';
import { Component, OnInit, Input } from '@angular/core';
import { Kriterien } from 'shared';
import { SafeResourceUrl, DomSanitizer } from '@angular/platform-browser';
import { map, tap } from 'rxjs/operators';

@Component({
  selector: 'inspector-kriterium',
  templateUrl: './level-inspector-kriterium.component.html',
  styleUrls: ['./level-inspector-kriterium.component.scss']
})
export class LevelInspectorKriteriumComponent implements OnInit {

  @Input() kriterium: Kriterien;
  @Input() isAdmin: boolean;
  imageData: SafeResourceUrl;

  constructor(
    private api: ApiService,
    private dom: DomSanitizer) { }

  ngOnInit() {
    this.loadImage();
  }

  loadImage() {
    if (this.kriterium.picto) {
      this.api.getPictoBase64(this.kriterium.picto.id, this.isAdmin)
        .pipe(
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
          tap(image => console.log('image', image)),
          tap(image => this.imageData = image),
        ).subscribe(() => {
          console.log('loaded image');
        }, err => console.log('couldnt load image', err));
    }
  }

}
