// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/task-kriterium/task-kriterium.component.ts
import { Component, OnInit, Input } from '@angular/core';
import { SafeResourceUrl, DomSanitizer } from '@angular/platform-browser';
import { Kriterien } from 'shared';
import { ApiService } from '@services/api.service';
import { map, tap } from 'rxjs/operators';

@Component({
  selector: 'app-task-kriterium',
  templateUrl: './task-kriterium.component.html',
  styleUrls: ['./task-kriterium.component.scss']
})
export class TaskKriteriumComponent implements OnInit {

  @Input() kriterium: Kriterien;
  imageData: SafeResourceUrl;

  constructor(public api: ApiService, private dom: DomSanitizer) { }

  ngOnInit() {
    this.loadImage();
  }

  loadImage() {
    if (this.kriterium.picto) {
      this.api.getPictoBase64(this.kriterium.picto.id)
        .pipe(
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
          tap(image => console.log('image', image)),
          tap(image => this.imageData = image),
        ).subscribe(() => {
          console.log('loaded image');
        }, err => console.log('couldnt load image', err));
    }
  }

}
