// @fivethree/billy-plugin-git/testi/management-app/src/app/components/checklist/checklist.component.ts
import { IonCheckbox } from '@ionic/angular';
import { Kriterien } from './../../../../../shared/types/types.d';
import { Component, OnInit, Input, EventEmitter, Output, ViewChildren, QueryList } from '@angular/core';
import { Checklist, Level } from 'shared';

@Component({
  selector: 'app-checklist',
  templateUrl: './checklist.component.html',
  styleUrls: ['./checklist.component.scss']
})
export class ChecklistComponent implements OnInit {

  @Input() checklist: Checklist;

  _selected: number[];
  @Input() first = false;
  @Input() last = false;

  @Output() addKriterien = new EventEmitter<Kriterien[]>();
  @Output() removeKriterien = new EventEmitter<Kriterien[]>();
  @ViewChildren(IonCheckbox) checkboxes = new QueryList<IonCheckbox>();

  constructor() { }

  ngOnInit() {

  }

  checklistChange(event) {


    const allKriterien = this.checklist.levels
      .map(level => level.kriterien)
      .reduce((a, b) => [...a, ...b]);

    if (event.detail.checked) {
      this.addKriterien.emit(allKriterien);
    } else {
      this.removeKriterien.emit(allKriterien);
    }
  }

  levelChange(event, level: Level) {

    if (event.detail.checked) {
      this.addKriterien.emit(level.kriterien);
    } else {
      this.removeKriterien.emit(level.kriterien);
    }
  }

}
