// @fivethree/billy-plugin-git/testi/management-app/src/app/components/components.module.ts
import { BreadcrumbComponent } from './breadcrumb/breadcrumb.component';
import { KriteriumSelectComponent } from './kriterium-select/kriterium-select.component';
import { ChecklistComponent } from './checklist/checklist.component';
import { MatBadgeModule } from '@angular/material/badge';
import { InspectionKriteriumComponent } from './inspection-kriterium/inspection-kriterium.component';
import { LevelComponent } from './level/level.component';
import { CommonModule } from '@angular/common';
import { IonicModule } from '@ionic/angular';
import { NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
import { UserPropertyActionsComponent } from './user-property-actions/user-property-actions.component';
import { UserPropertyAddComponent } from './user-property-add/user-property-add.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { PropertyChecklistAddComponent } from './property-checklist-add/property-checklist-add.component';
import { PropertyChecklistActionsComponent } from './property-checklist-actions/property-checklist-actions.component';
import { FivethreeCoreModule } from '@fivethree/core';
import { KriteriumComponent } from './kriterium/kriterium.component';
import { LevelActionsComponent } from './level-actions/level-actions.component';
import { LevelInspectorComponent } from './level-inspector/level-inspector.component';
import { LevelInspectorKriteriumComponent } from './level-inspector-kriterium/level-inspector-kriterium.component';
import { MatTooltipModule } from '@angular/material/tooltip';
import { MatTableModule } from '@angular/material/table';
import { MatPaginatorModule } from '@angular/material/paginator';
import { MatSortModule } from '@angular/material/sort';
import { MatSelectModule } from '@angular/material/select';
import { MatTabsModule } from '@angular/material/tabs';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { EditableComponent } from './editable/editable.component';
import { MatMomentDateModule } from '@angular/material-moment-adapter';
import { InspectionResultComponent } from './inspection-result/inspection-result.component';
import { PipesModule } from '@pipes/pipes.module';
import { MatFormFieldModule, MatInputModule } from '@angular/material';
import { PictoComponent } from './picto/picto.component';
import { PictosPage } from './pictos/pictos.page';
import { TaskCreateComponent } from './task-create/task-create.component';
import { TaskKriteriumComponent } from './task-kriterium/task-kriterium.component';

@NgModule({
    imports: [
        FivethreeCoreModule,
        MatIconModule,
        MatTableModule,
        IonicModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        MatTooltipModule,
        MatPaginatorModule,
        MatSortModule,
        MatSelectModule,
        MatTabsModule,
        MatBadgeModule,
        PipesModule,
        MatDatepickerModule,
        MatMomentDateModule,
        MatFormFieldModule,
        MatInputModule
    ],
    exports: [
        MatTooltipModule,
        MatTableModule,
        FivethreeCoreModule,
        MatIconModule,
        CommonModule,
        LevelComponent,
        FormsModule,
        ReactiveFormsModule,
        LevelInspectorComponent,
        LevelInspectorKriteriumComponent,
        UserPropertyActionsComponent,
        UserPropertyAddComponent,
        LevelComponent,
        PropertyChecklistAddComponent,
        PropertyChecklistActionsComponent,
        KriteriumComponent,
        LevelActionsComponent,
        MatPaginatorModule,
        MatSortModule,
        MatSelectModule,
        EditableComponent,
        InspectionKriteriumComponent,
        InspectionResultComponent,
        ChecklistComponent,
        KriteriumSelectComponent,
        MatTabsModule,
        MatDatepickerModule,
        MatMomentDateModule,
        MatFormFieldModule,
        MatInputModule,
        PictoComponent,
        TaskCreateComponent,
        BreadcrumbComponent,
        TaskKriteriumComponent,
        MatBadgeModule],
    declarations: [
        LevelInspectorComponent,
        LevelInspectorKriteriumComponent,
        UserPropertyActionsComponent,
        UserPropertyAddComponent,
        LevelComponent,
        PropertyChecklistAddComponent,
        PropertyChecklistActionsComponent,
        KriteriumComponent,
        LevelActionsComponent,
        InspectionKriteriumComponent,
        InspectionResultComponent,
        EditableComponent,
        ChecklistComponent,
        KriteriumSelectComponent,
        PictosPage,
        PictoComponent,
        TaskCreateComponent,
        BreadcrumbComponent,
        TaskKriteriumComponent],
    entryComponents: [
        UserPropertyActionsComponent,
        UserPropertyAddComponent,
        PropertyChecklistAddComponent,
        PropertyChecklistActionsComponent,
        LevelActionsComponent,
        PictosPage
    ],
    providers: [],
})
export class ComponentsModule { }
