// @fivethree/billy-plugin-git/immoapps/management-app/src/app/pages/password/password.page.ts
import { FivLoadingButton } from '@fivethree/core';
import { NavController, IonInput } from '@ionic/angular';
import { Component, OnInit, ViewChild } from '@angular/core';
import { FormBuilder, FormGroup, FormControl, Validators } from '@angular/forms';
import { AuthService } from '../../services/auth.service';
import { ToastService } from '../../services/toast.service';

@Component({
  selector: 'app-password',
  templateUrl: './password.page.html',
  styleUrls: ['./password.page.scss'],
})
export class PasswordPage {

  password: FormGroup;
  @ViewChild('first') passwordField: IonInput;

  constructor(public formBuilder: FormBuilder, public auth: AuthService, public navCtrl: NavController,
    public toast: ToastService) {
    this.password = formBuilder.group({
      password1: new FormControl('', [Validators.minLength(6), Validators.required]),
      password2: new FormControl('', [Validators.minLength(6), Validators.required]),
    });
  }

  ionViewWillEnter() {
  }

  changePassword(loading: FivLoadingButton) {
    loading.load();
    this.auth.changePassword({ newPassword: this.password.value.password1 }).subscribe(() => {
      this.toast.presentToast(8000, 'Passwort erfolgreich geändert.', 'OK');
      loading.complete();
    }, err => {
      this.onError();
      loading.unload();
    });
  }
  onError(): any {
    this.toast.presentToast(5000, 'Ein unerwarteter Fehler ist aufgetreten. Bitte überprüfen Sie Ihre Eingabe und versuchen Sie es erneut.', 'OK').then(toast => {
      if (toast.role === 'cancel') {
        this.passwordField.setFocus();
      }
    });
  }



  changePasswordComplete() {
    this.navCtrl.back();
  }
}
