// @fivethree/billy-plugin-git/testi/management-app/src/app/components/property-checklist-add/property-checklist-add.component.ts
import { Checklist, Property } from 'shared';
import { Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';
import { DataService } from '../../services/data.service';
import { NavParams } from '@ionic/angular';
import { map } from 'rxjs/operators';

@Component({
  selector: 'app-property-checklist-add',
  templateUrl: './property-checklist-add.component.html',
  styleUrls: ['./property-checklist-add.component.scss']
})
export class PropertyChecklistAddComponent implements OnInit {

  checklists: Observable<Checklist[]>;

  constructor(public data: DataService, public navParams: NavParams) { }

  ngOnInit() {
    const propertyChecklists: Checklist[] = this.navParams.get('propertyChecklists');

    this.checklists = this.data.loadChecklists()
      .pipe(
        map(checklists => checklists.filter(checklist => !propertyChecklists.find(c => c.id == checklist.id)))
      );
  }

  addChecklist(checklist: Checklist) {
    this.navParams.get('controller').dismiss(checklist, 'add');
  }

}
