// @fivethree/billy-plugin-git/testi/management-app/src/app/pages/password-change/password-change.page.ts
import { FivLoadingProgressBar } from '@fivethree/core';
import { Component } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from '@angular/forms';
import { Router } from '@angular/router';
import { ChangePasswordPayload } from 'shared';
import { ApiService } from '@services/api.service';
import { ToastService } from '@services/toast.service';

@Component({
  selector: 'app-password-change',
  templateUrl: './password-change.page.html',
  styleUrls: ['./password-change.page.scss'],
})
export class PasswordChangePage {

  passwordChangeForm: FormGroup;
  constructor(private router: Router,
    private formBuilder: FormBuilder,
    private api: ApiService,
    public toast: ToastService) {
    this.setupForm();
  }

  ionViewWillEnter() {
  }

  setupForm() {
    this.passwordChangeForm = this.formBuilder.group({
      newPassword: ['', [Validators.minLength(6), Validators.required]],
    });
  }

  changePassword(loading: FivLoadingProgressBar) {
    loading.load();
    const payload: ChangePasswordPayload = this.passwordChangeForm.value;
    this.api.changePassword(payload)
      .subscribe(() => loading.complete({}),
        (error) => {
          this.onError();
          loading.unload();
        });
  }

  passwordChangeComplete() {
    this.router.navigate(['/settings']);
  }

  onError(): any {
    this.toast.presentToast(5000, 'Ein unerwarteter Fehler ist aufgetreten. Bitte überprüfen Sie Ihre Eingabe und versuchen Sie es erneut.', 'OK');
  }
}
