// @fivethree/billy-plugin-git/testi/management-app/src/app/services/toast.service.ts
import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

@Injectable({
  providedIn: 'root'
})
export class ToastService {

  currentToast;

  constructor(public toastController: ToastController) { }

  async presentToast(duration: number, message: string, action: string) {
    if (this.currentToast) {
      this.currentToast.dismiss();
    }
    this.currentToast = await this.toastController.create({
      message: message,
      duration: duration,
      closeButtonText: action,
      showCloseButton: true
    });

    await this.currentToast.present();
    return await this.currentToast.onDidDismiss();
  }

  async success(duration: number, message: string, action: string) {
    if (this.currentToast) {
      this.currentToast.dismiss();
    }
    this.currentToast = await this.toastController.create({
      message: message,
      duration: duration,
      closeButtonText: action,
      showCloseButton: true,
      cssClass: 'toast-success'
    });

    await this.currentToast.present();
    return await this.currentToast.onDidDismiss();
  }

  async error(duration: number, message: string, action: string) {
    if (this.currentToast) {
      this.currentToast.dismiss();
    }
    this.currentToast = await this.toastController.create({
      message: message,
      duration: duration,
      closeButtonText: action,
      showCloseButton: true,
      cssClass: 'toast-danger'
    });

    await this.currentToast.present();
    return await this.currentToast.onDidDismiss();
  }

  async warning(duration: number, message: string, action: string) {
    if (this.currentToast) {
      this.currentToast.dismiss();
    }
    this.currentToast = await this.toastController.create({
      message: message,
      duration: duration,
      closeButtonText: action,
      showCloseButton: true,
      cssClass: 'toast-warning'
    });

    await this.currentToast.present();
    return await this.currentToast.onDidDismiss();
  }
}
