// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/pages/check/check.page.ts
import { ContextState } from './../../store/context/context.state';
import { NavigateBackward } from '@fivethree/ngxs-ionic-router-plugin';
import { ActivatedRoute } from '@angular/router';
import { PropertiesState } from '@store/properties/properties.state';
import { Component, OnInit, ViewChild, ViewChildren, QueryList, AfterViewInit } from '@angular/core';
import { Property, Kriterien, AddInspectionPayload, GetContextResult } from 'shared';
import { Observable } from 'rxjs';
import { FivStep, FivStepper, DrawerState, FivLoadingProgressBar } from '@fivethree/core';
import { SelectContentComponent } from '../../components/select-content/select-content.component';
import { Select, Store } from '@ngxs/store';
import { NotifyHint, NotifyError } from '@store/notify/notify.actions';
import { CreateInspection } from '@store/inspections/inspections.actions';
import { Dispatch } from '@ngxs-labs/dispatch-decorator';
import { GetContext } from '@store/context/context.actions';

@Component({
  selector: 'app-check',
  templateUrl: './check.page.html',
  styleUrls: ['./check.page.scss'],
})
export class CheckPage implements OnInit, AfterViewInit {

  @Select(PropertiesState.properties)
  allProperties: Observable<Property[]>;

  property: Property;
  kriterien: Kriterien[] = [];

  drawerState: DrawerState = DrawerState.Bottom;
  @ViewChild(FivStepper) stepper: FivStepper;
  @ViewChildren(FivStep) steps: QueryList<FivStep>;

  @ViewChild(SelectContentComponent) selectContent: SelectContentComponent;

  propertySelectOptions: any = {
    header: 'Immobilie ausw채hlen'
  };

  constructor(private store: Store,
    private route: ActivatedRoute) { }

  ngOnInit() {
    const propertyId = +this.route.snapshot.queryParamMap.get('id');
    if (propertyId) {
      this.property = this.store.selectSnapshot(PropertiesState.getProperty(propertyId));
    }
  }

  ngAfterViewInit() {
    this.stepper.open(0);
  }

  ionViewDidEnter() {
    if (this.property) {
      this.stepper.close(0);
      this.stepper.completeStep(0);
      this.stepper.open(1);
    }
  }

  onStepClicked(index: number) {
    switch (index) {
      case 0: this.stepper.open(index); break;
      case 1: this.checksStepClicked(); break;
    }
  }

  checksStepClicked() {
    if (this.property) {
      this.stepper.open(1);
    } else {
      this.noPropertyError();
    }
  }

  async noPropertyError() {
    this.store.dispatch(new NotifyHint('Bitte w채hlen Sie zun채chst eine Immobilie aus'));
    if (!this.steps.toArray()[0].open) {
      this.stepper.open(0);
    }
  }

  selectProperty(event) {
    if (!event.detail || !event.detail.value) {
      return;
    }
    this.kriterien = [];
    this.property = event.detail.value;
    this.stepper.close(0);
    this.stepper.completeStep(0);
    if (this.selectContent) {
      this.selectContent.property = this.property;
      this.selectContent.load();
      this.selectContent.reset();
    }

    setTimeout(() => {
      this.stepper.open(1);
    }, 250);
  }

  addCriteria() {
    this.drawerState = DrawerState.Top;
  }

  onDone() {
    this.selectContent.onDone();
  }
  onAbort() {
    this.selectContent.onAbort();
  }

  abort() {
    this.drawerState = DrawerState.Bottom;
  }

  onCriteriaSelected(kriterien: Kriterien[]) {
    console.log('selected kriterien', kriterien);
    this.kriterien = kriterien;
    this.drawerState = DrawerState.Bottom;
  }

  changeSelection() {
    this.selectContent.set(this.kriterien);
    this.addCriteria();
  }

  addInspection(bar: FivLoadingProgressBar) {
    bar.load();
    const payload: AddInspectionPayload = {
      propertyId: this.property.id,
      inspectionCriterias: this.kriterien.map(kriterium => kriterium.id)
    };


    this.store.dispatch(new CreateInspection(payload))
      .subscribe((inspection) => bar.complete(inspection),
        err => {
          this.store.dispatch(new NotifyError('Die Begehung konnte nicht angelegt werden. Bitte versuchen Sie es sp채ter erneut.', err));
          bar.unload();
        });
  }

  @Dispatch()
  onSuccess = () => new NavigateBackward(['tabs', 'tab2'])

}
