// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/picto/picto.component.ts
import { AlertController } from '@ionic/angular';
import { Component, OnInit, Input, HostListener, ElementRef, OnDestroy, EventEmitter, Output } from '@angular/core';
import { GetPictoBase64Result } from 'shared';
import { Subscription, fromEvent } from 'rxjs';
import { debounceTime, takeUntil, repeat } from 'rxjs/operators';

@Component({
  selector: 'app-picto',
  templateUrl: './picto.component.html',
  styleUrls: ['./picto.component.scss']
})
export class PictoComponent implements OnInit, OnDestroy {


  @Input() picto: GetPictoBase64Result;
  hovering = false;

  enter: Subscription;
  leave: Subscription;

  @Output() selectPicto = new EventEmitter<GetPictoBase64Result>();
  @Output() deletePicto = new EventEmitter<GetPictoBase64Result>();

  constructor(private element: ElementRef, private alertController: AlertController) {
    this.enter = fromEvent(element.nativeElement, 'mouseenter')
      .pipe(
        debounceTime(1000),
        takeUntil(fromEvent(element.nativeElement, 'mouseleave')),
        repeat()
      ).subscribe(() => this.hovering = true);

    this.leave = fromEvent(element.nativeElement, 'mouseleave')
      .subscribe(() => this.hovering = false);
  }

  ngOnInit() {
  }

  ngOnDestroy(): void {
    this.enter.unsubscribe();
    this.leave.unsubscribe();
  }

  delete() {
    this.deletePicto.emit(this.picto);
  }

  async presentAlertConfirm() {
    const alert = await this.alertController.create({
      header: 'Löschen bestätigen',
      message: 'Symbolbild wirklich löschen?',
      buttons: [
        {
          text: 'Abbrechen',
          role: 'cancel',
          cssClass: 'secondary'
        }, {
          text: 'Löschen',
          handler: () => {
            this.delete();
          }
        }
      ]
    });
  
    await alert.present();
  }

  select() {
    this.selectPicto.emit(this.picto);
  }

}
