// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/pages/user/user.page.ts
import { NavigateForward } from '@fivethree/ngxs-ionic-router-plugin';
import { Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';
import { User } from 'shared';
import { Select } from '@ngxs/store';
import { Dispatch } from '@ngxs-labs/dispatch-decorator';
import { Logout } from '@store/auth/auth.actions';
import { AuthState } from '@store/auth/auth.state';
import { NotifyHint } from '@store/notify/notify.actions';

@Component({
  selector: 'app-user',
  templateUrl: './user.page.html',
  styleUrls: ['./user.page.scss'],
})
export class UserPage implements OnInit {

  @Select(AuthState.user)
  user: Observable<User>;

  constructor() { }

  ngOnInit() {
  }

  @Dispatch()
  signOut = () => new Logout()

  @Dispatch()
  presentToast = () => new NotifyHint('Diese Funktion wird gerade entwickelt.')

  @Dispatch()
  changePassword = () => new NavigateForward(['password/change'])

  @Dispatch()
  imprint = () => new NavigateForward(['imprint'])

  @Dispatch()
  privacy = () => new NavigateForward(['privacy'])

  @Dispatch()
  userDetail = () => new NavigateForward(['user/detail'])

}
