// @fivethree/billy-plugin-git/testi/immoapp/src/app/components/inspection-result/inspection-result.component.ts
import { HttpClient } from '@angular/common/http';
import { DomSanitizer, SafeResourceUrl } from '@angular/platform-browser';
import { ApiService } from '@services/api.service';
import { Component, OnInit, Input } from '@angular/core';
import { Result, Image } from 'shared';
import { map, tap } from 'rxjs/operators';

@Component({
  selector: 'app-inspection-result',
  templateUrl: './inspection-result.component.html',
  styleUrls: ['./inspection-result.component.scss']
})
export class InspectionResultComponent implements OnInit {
  @Input() result: Result;
  image: SafeResourceUrl;

  constructor(private api: ApiService, private dom: DomSanitizer, private http: HttpClient) { }

  ngOnInit(): void {
    if (this.result.images && this.result.images[0]) {
      this.http.get<Image>(this.result.images[0].url)
        .pipe(
          tap(image => { console.log('####', image); }),
          map(image => this.dom.bypassSecurityTrustResourceUrl(`data:${image.type};base64,${image.data}`)),
        )
        .subscribe(image => {
          this.image = image;
        }, err => {
          console.log(err);
        });
    }
  }



}
