// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/tabs/tabs.page.ts
import { Component, OnInit } from '@angular/core';
import { NavigateRoot, NavigateForward } from '@fivethree/ngxs-ionic-router-plugin';
import { Dispatch } from '@ngxs-labs/dispatch-decorator';
import { Store } from '@ngxs/store';
import { InspectionsState } from '@store/inspections/inspections.state';
import { Observable } from 'rxjs';
import { MessagesState } from '@store/messages/messages.state';

@Component({
  selector: 'app-tabs',
  templateUrl: './tabs.page.html',
  styleUrls: ['./tabs.page.scss'],
})
export class TabsPage implements OnInit {

  inspectionsCount: Observable<number>;
  unreadNotificationCount: Observable<number>;

  constructor(private store: Store) { }

  ngOnInit() {
    this.inspectionsCount = this.store.select(InspectionsState.inspectionsCount);
    this.unreadNotificationCount = this.store.select(MessagesState.unreadNotificationCount);
  }

  @Dispatch()
  onClick = () => new NavigateRoot('/tabs/(tab2:tab2)')

  @Dispatch()
  startCheck = () => new NavigateForward('/check')

}
