// @fivethree/billy-plugin-git/immoapps/immoapp/src/app/components/components.module.ts
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { KriteriumComponent } from './kriterium/kriterium.component';
import { PipesModule } from './../pipes/pipes.module';
import { CommonModule } from '@angular/common';
import { FivethreeCoreModule } from '@fivethree/core';
import { IonicModule } from '@ionic/angular';
import { NgModule } from '@angular/core';
import { PropertyComponent } from './property/property.component';
import { ChecklistComponent } from './checklist/checklist.component';
import { MatTabsModule } from '@angular/material/tabs';
import { LevelComponent } from './level/level.component';
import { MatBadgeModule } from '@angular/material/badge';
import { SelectContentComponent } from './select-content/select-content.component';
import { InspectionComponent } from './inspection/inspection.component';
import { InspectKriteriumComponent } from './inspect-kriterium/inspect-kriterium.component';
import { MessageComponent } from './message/message.component';
import { InspectionResultComponent } from './inspection-result/inspection-result.component';



@NgModule({
    imports: [IonicModule, FivethreeCoreModule, CommonModule, PipesModule, MatTabsModule, MatBadgeModule,
        FormsModule, ReactiveFormsModule
    ],
    exports: [PropertyComponent, ChecklistComponent, KriteriumComponent, FivethreeCoreModule,
        InspectionComponent, InspectKriteriumComponent, FormsModule, ReactiveFormsModule,
        PipesModule, MatTabsModule, LevelComponent, MatBadgeModule, SelectContentComponent, MessageComponent, InspectionResultComponent],
    declarations: [PropertyComponent, ChecklistComponent, KriteriumComponent, LevelComponent,
        SelectContentComponent, InspectionComponent, InspectKriteriumComponent, MessageComponent, InspectionResultComponent],
    entryComponents: [],
    providers: [],
})
export class ComponentsModule { }
