// @fivethree/billy-plugin-git/immoapps/management-app/src/app/components/level-inspector/level-inspector.component.ts
import { Component, OnInit, Input, AfterContentInit, Output, EventEmitter } from '@angular/core';
import { Checklist } from 'shared';

@Component({
  selector: 'level-inspector',
  templateUrl: './level-inspector.component.html',
  styleUrls: ['./level-inspector.component.scss']
})
export class LevelInspectorComponent implements OnInit, AfterContentInit {

  @Input() checklist: Checklist;
  @Input() isAdmin: boolean;
  @Input() level: number;
  @Input() first = false;
  @Input() last = false;
  vals: boolean[] = [];

  @Output() next = new EventEmitter<number>();
  @Output() previous = new EventEmitter<number>();
  @Output() complete = new EventEmitter<number>();

  constructor() { }

  ngOnInit() {
  }

  ngAfterContentInit(): void {
    this.checklist.levels[this.level].kriterien.forEach(k => {
      this.vals.push(false);
    });
  }

  isComplete() {
    let valid = true;
    this.vals.forEach(val => {
      if (!val) {
        valid = false;
      }
    });

    return valid;
  }

}
