// @fivethree/billy-plugin-git/testi/immoapp/src/app/pages/password-change/password-change.page.ts
import { FivLoadingProgressBar } from '@fivethree/core';
import { OnInit, Component } from '@angular/core';
import { FormGroup, FormBuilder, FormControl, Validators, AbstractControl } from '@angular/forms';
import { ApiService } from '@services/api.service';
import { NavController } from '@ionic/angular';
import { NotifyService } from '@services/notify.service';
import { ChangePasswordPayload } from 'shared';
import { Dispatch } from '@ngxs-labs/dispatch-decorator';
import { NavigateBackward } from '@fivethree/ngxs-ionic-router-plugin';
@Component({
  selector: 'app-password-change',
  templateUrl: './password-change.page.html',
  styleUrls: ['./password-change.page.scss'],
})
export class PasswordChangePage implements OnInit {

  passwordChange: FormGroup;

  constructor(public formbuilder: FormBuilder,
    private api: ApiService,
    private navCtrl: NavController,
    private toast: NotifyService) {
    this.passwordChange = formbuilder.group({
      password: new FormControl('', [Validators.required]),
      password2: new FormControl('', [Validators.required]),
    });
  }

  ngOnInit() {
  }

  get password(): AbstractControl {
    return this.passwordChange.get('password');
  }
  get password2(): AbstractControl {
    return this.passwordChange.get('password2');
  }

  equal() {
    return this.password.value === this.password2.value;
  }

  change(loading: FivLoadingProgressBar) {
    loading.load();
    const payload: ChangePasswordPayload = {
      newPassword: this.password.value
    };
    this.api.changePassword(payload)
      .subscribe(() => {
        loading.complete({});
        this.toast.success('Ihr Passwort wurde geändert.');
      }, err => {
        this.toast.error('Ein Fehler ist aufgetreten. Bitte versuchen Sie es später erneut');
        loading.unload();
      });
  }

  changeComplete() {
    this.navCtrl.back();
  }

  @Dispatch()
  navigateBack = () => new NavigateBackward(['tabs','tab4'])

}
